src/Features/Core/Portable/InlineMethod/AbstractInlineMethodRefactoringProvider.MethodParametersInfo.cs

```diff
@@ -0,0 +1,297 @@
+            /// </summary>
+            public ImmutableArray<IParameterSymbol> ParametersWithDefaultValue { get; }
+
+            /// <summary>
+            /// Parameters map to literal expression argument.
+            /// </summary>
+            public ImmutableArray<(IParameterSymbol parameterSymbol, SyntaxNode literalExpressionSyntaxNode)> ParametersWithLiteralArgument { get; }
+
+            private MethodParametersInfo(
+                ImmutableArray<(IParameterSymbol, SyntaxNode)> parametersWithIdentifierArgument,
+                ImmutableArray<(IParameterSymbol parameterSymbol, SyntaxNode variableDeclarationNode)> parametersWithVariableDeclarationArgument,
+                ImmutableArray<(IParameterSymbol, ImmutableArray<SyntaxNode>)> parametersNeedGenerateDeclarations,
+                ImmutableArray<IParameterSymbol> parametersWithDefaultValue,
+                ImmutableArray<(IParameterSymbol parameterSymbol, SyntaxNode literalExpressionSyntaxNode)> parametersWithLiteralArgument)
+            {
+                ParametersWithIdentifierArgument = parametersWithIdentifierArgument;
+                ParametersWithVariableDeclarationArgument = parametersWithVariableDeclarationArgument;
+                ParametersNeedGenerateDeclarations = parametersNeedGenerateDeclarations;
+                ParametersWithDefaultValue = parametersWithDefaultValue;
+                ParametersWithLiteralArgument = parametersWithLiteralArgument;
+            }
+
+            public static MethodParametersInfo GetMethodParametersInfo(
+                AbstractInlineMethodRefactoringProvider<TInvocationSyntaxNode, TExpressionSyntax, TArgumentSyntax> inlineMethodRefactoringProvider,
+                ISyntaxFacts syntaxFacts,
+                SemanticModel semanticModel,
+                SyntaxNode calleeInvocationSyntaxNode,
+                IMethodSymbol calleeMethodSymbol,
+                CancellationToken cancellationToken)
+            {
+                // Classify the parameters into:
+                // 1. If the parameter accept an identifier from Caller, then use this identifier to replace all the occurence of the parameter.
+                // (So that after inlining, the same identifier could be found in the caller)
+                // 2. If the parameter accept literal/no argument but has default value, replace all the occurence of the parameter by using the
+                // literal expression
+                // 3. If the argument has variable declarations ('out var' in C#), then insert an additional declaration. Also replace all the
+                // occurence of parameter by using this identifier.
+                // 4. For the rest of the parameters (might include method invocation and different expressions), use the parameter's name to
+                // generate a declaration for them and insert that into caller.
+                var allArguments = syntaxFacts.GetArgumentsOfInvocationExpression(calleeInvocationSyntaxNode);
```

Comment:
general question: do you have tests that inline an extension method.

Reply:
I think the design is only support private method. So I haven't consider the extension method cases & overloading operator because I think they need to be internal at least

Reply:
extension methods can be private :)

Reply:
Ah, I tested a sample code and it could be private. Will make sure there is a test case for that

Reply:
Done, also add test to cover extension method